### DISCUSSION TITLE
Client configuration for self hosted server

### DISCUSSION THREAD

**Discussion by ghost** [rustdesk/rustdesk#966] [Q&A] [answered]
For Windows you are able to configure the host and the public key via the filename, which looks very suspicious.

`rustdesk-host=MY.DOMAIN,key=MY-PUBLIC-KEY=.exe`

It would be much more comfortable for a user to be able to use a configuration instead of this cryptic file name.

Possible solutions:

1. Use a config-file: Check if a configuration file exists and if so, use the contents for HOST and KEY (JSON, XML, Plain Text)
2. Add parameters: e. g. -h and -k to be able to use a start.bat (Windows) or shell script (Linux) to call the executable with these parameters.

Best regards

**Comment by plittlefield** [comment]
I second this, good idea.  The documented way with the host and public key in the filename just doesn't work for me :-(  It seems to ignore it!

**Reply by ghost** [reply]
Make sure to have only valid Windows filename characters in your public key. I had problems first too because I had an "/" in my public key. If that's the case for you:

1. delete id_ed25519.pub und id_ed25519 in ~/rustdesk/hbbs
2. restart container or what kind of installation you have
3. repeat until you have only valid characters for Windows filenames in your public key

**Reply by plittlefield** [reply]
My public key only has regular characters in it APART from the = equals sign at the end... is this OK for Windows?

'rustdesk-host=rustdesk.mydomain.uk,key=KjPK1Q7t3xxxxxxxxxxxxxxxxxxxxxxmi7DlwQ=.exe'

**Reply by ghost** [reply]
If the xxxx are normal characters, it looks finde for me. The "=" is no problem.

Forbidden characters are: 
`~ “ # % & * : < > ? / \ { | }`

**Reply by oboudry-mvp** [reply]
It worked for me but I had to add `--silent-install` when running the renamed exe. I agree it's a very cumbersome way of configuring the app.

**Comment by nemiah** [comment]
I am also looking for a way to provide clients for all the supported platforms with my own server settings. Probably the only feasible way is to compile the clients for all the platforms by myself. Is someone here interested on working with me on this?

**Reply by plittlefield** [reply]
I'll give it a go.

However, IMHO ...  if the developers can fix the clever "filename trick"  (as I'm calling it) for all platforms, then I think that's the way to go.

If not, then a .cfg alongside the .exe is second best.

**Reply by nemiah** [reply]
The filename is fine if you use the files yourself, I agree.
I want to use the client to support the users of my software and the filename-approach does not work for me in this case.
A cfg-file makes it unnecessarily complicated, it's hard to explain via phone.
A single .exe(/.dmg/.AppImage-file) just has the lowest friction.

**Reply by mpember** [reply]
It would be great if there was some sort of build tool that allows for the creation of a client app that embeds the relevant key/host details.

While testing, I have opted to use a custom URL shortener to keep a simple URL for end users to download the client.

Using the filename method causes issues when an end user is unaware that they have already downloaded the file.  Because most browsers will simply append something to the end of the filename, it breaks the key.

**Reply by nemiah** [reply]
Yes, some sort of generator is my idea, too. It would be great for Mac, Win and Linux at least.
I can already build the client for Linux but I'm not getting the .deb-file as the build.py-script is not working for me yet.
Cross-compiling from Linux to Mac and Win would be ideal, but I'm not sure if this will work.
Do you know how the release-files (deb/AppImage, dmg, exe) are created?

**Reply by Tacioandrade** [reply]
> The filename is fine if you use the files yourself, I agree. I want to use the client to support the users of my software and the filename-approach does not work for me in this case. A cfg-file makes it unnecessarily complicated, it's hard to explain via phone. A single .exe(/.dmg/.AppImage-file) just has the lowest friction.

In your case it would not be necessary to explain over the phone, you can use winrar, winzip or 7zip to generate an executable with the rustdesk exe, along with the configuration file.
It's the same logic as the pfSense OpenVPN installer, for example, which is generated at runtime by pfSense itself by zipping the ovpn configuration file + openvpn installer.

**Reply by Tacioandrade** [reply]
> It would be great if there was some sort of build tool that allows for the creation of a client app that embeds the relevant key/host details.
> 
> While testing, I have opted to use a custom URL shortener to keep a simple URL for end users to download the client.
> 
> Using the filename method causes issues when an end user is unaware that they have already downloaded the file. Because most browsers will simply append something to the end of the filename, it breaks the key.

If you store on your own web server with apache, it would be simple, just create a https://multiti.com.br/rustdesk.php for example and this .php page forces you to download the file with the full extension with a redirect to for example https://multiti.com.br/files/rustdesk-host=MY.DOMAIN,key=MY-PUBLIC-KEY=.exe
The content of rustdesk.php could be:
<?php
header('Location: https://multiti.com.br/files/rustdesk-host=MY.DOMAIN,key=MY-PUBLIC-KEY=.exe');
?>

**Reply by mpember** [reply]
There is no need for PHP code.  You could do it with a simple Apache redirect.  In my case, I have a personal URL shortener that turns https://mydomain.com/remote into https://files.mydomain.com/files/rustdesk-host=MY.DOMAIN,key=MY-PUBLIC-KEY=.exe

But that doesn't change the problem that the key gets mangled when it is accidentally downloaded a second time.

It also offers no solution for Mac.

**Reply by kintaro1981** [reply]
> > The filename is fine if you use the files yourself, I agree. I want to use the client to support the users of my software and the filename-approach does not work for me in this case. A cfg-file makes it unnecessarily complicated, it's hard to explain via phone. A single .exe(/.dmg/.AppImage-file) just has the lowest friction.
> 
> In your case it would not be necessary to explain over the phone, you can use winrar, winzip or 7zip to generate an executable with the rustdesk exe, along with the configuration file. It's the same logic as the pfSense OpenVPN installer, for example, which is generated at runtime by pfSense itself by zipping the ovpn configuration file + openvpn installer.

Why do you need a cfg file? You can simply 7zip the "long name exe" generating a generic my_rustdesk.exe and execute rustdesk-host=MY.DOMAIN,key=MY-PUBLIC-KEY=.exe from it. If it works the MS Windows part is solved with this workaround.

**Comment by zodiacg** [comment]
Second this. 

My selfhosted server used non-standard port and I can't write it in the filename since ":" is not usable for filename in Windows. PubKey will have similar problem if non-usable characters are used in the key.

-h/-k parameters are common implementation. Though I think "config in the filename" is a very useful feature in cases a single executable is more preferred. I would suggest using an escaped/encoded string to avoid the non-usable character problem.

**Reply by rustdesk** [reply]
https://github.com/rustdesk/rustdesk-server/commit/848b5aedb7e8bda0f582706c792ffbab2d657a39

**Reply by zodiacg** [reply]
That's some hyper speed response :)

**Comment by pixelblut** [comment]
IMHO a config file (config.txt) at the same place as the .exe or under /etc shouldn't be that hard to implement and can easily be deployed to Clients via Ansible, SCCM or OS tools like scp/rsync/PowerShell.

My Rust isn't that good (to be fair, it's sysengineer-hacky-just-bend-it-until-it-works-horrible), but a quick google-fu and searching the Rust docs returned that lines can be iterated by [read_lines](https://doc.rust-lang.org/rust-by-example/std_misc/file/read_lines.html) from a file, which theoretically only needs to be run once when the Client starts. 
And all that's really needed in this config file are the params we have to put into the filename of the binary at the moment and pass it internally like the params are handled now.

Another simple method would be to read [ENV variables](https://doc.rust-lang.org/book/ch12-05-working-with-environment-variables.html), which theoretically is even more os agnostic than a config file. Wouldn't be that hard to implement, as the current Version already reads env variables for all kind of stuff like DISPLAY, XDG_CURRENT_DESKTOP, SystemDrive or ProgramFiles.

**Comment by goodluck1982** [comment]
This is just what I suggested in https://github.com/rustdesk/rustdesk/discussions/1873#discussion-4520950 [the (3) point].

**Reply by pedjas** [reply]
In old times, it was not complicated to add custom data at the end of exe, and exe was able to access it. If that technique is feasible with modern tools that cold be way to solve an issue.

In that case user could start RustDesk client, set it up, including host, port, key and whatever else) and then click option to produce new exe with all that settings attached. New exe, when run, would red settings from itself and reconfigure.

**Comment by elico** [comment]
> For Windows you are able to configure the host and the public key via the filename, which looks very suspicious.
> 
> `rustdesk-host=MY.DOMAIN,key=MY-PUBLIC-KEY=.exe`
> 
> It would be much more comfortable for a user to be able to use a configuration instead of this cryptic file name.
> 
> Possible solutions:
> 
>     1. Use a config-file: Check if a configuration file exists and if so, use the contents for HOST and KEY (JSON, XML, Plain Text)
> 
>     2. Add parameters: e. g. -h and -k to be able to use a start.bat (Windows) or shell script (Linux) to call the executable with these parameters.
> 
> 
> Best regards

Was there any progress  with this?
I have tested this format with my installation but wasn't able to use this format:
`rustdesk-host=ngtech.co.il,key=XYZ=.exe`

I took the installer and changed it's name.
Then when I start the software it picks the settings from the toml file that exist in the user/service directory.
What are the conditions for this file format to work?
I also want to pack the binary inside an archive so I can start it in such a way that it will 100% pick the new config.
What would be recommended for that?
Can we write a config host and key updater?

**Reply by plittlefield** [reply]
Honestly, a file called rustdesk.cfg alongside the .exe would work.  That way, I can get my client to download both files and run the .exe so it picks up the config.

**Reply by pedjas** [reply]
Remove installed version and use self executing exe. That is only way that standalone renamed to connect to self hosted server would work.

Check https://github.com/rustdesk/rustdesk/discussions/2152

**Reply by elico** [reply]
@pedjas Then what is needed now is probably an embedded "local config" cleanup or "uninstall" button.
To actually remove rustdesk local service configuration we need to be... admin user.
If the client is not an admin user then we need a proper option for that to work.
For example a "fully portable" instance of rustdesk is one way.
If the user is admin then we need to cleanup any:
`RustDesk.toml`
and
`RustDesk2.toml`

(on windows)
from the user roaming appdata folder ie:
`%AppData%\Roaming\RustDesk\config`

And also local services roaming profile which is at under the windows folder (for example):
`C:\Windows\ServiceProfiles\LocalService\AppData\Roaming\RustDesk\`

About the serf extracting exe it's pretty simple to do.
The other issue is the cleanup.
If the rustdesk project or anyone else can help to write a cleanup and uninstaller script it would help a lot.
The uninstaller can be based on the next script:
https://github.com/techahold/rustdeskinstall/blob/master/WindowsAgentAIOInstall.ps1

which  works great on windows 10 and up.

I have also found another issue with the file naming.
There should be an option to use another parameter in the file name after a ",".
Currently the only way it works is if the file name is constructed with only the `host=hostname,key=key.exe`
I would have expected for it to support also `host=hostname,key=key,company=x,otherkey=value.exe`

**Reply by pedjas** [reply]
> @pedjas Then what is needed now is probably an embedded "local config" cleanup or "uninstall" button. To actually remove rustdesk local service configuration we need to be... admin user. If the client is not an admin user then we need a proper option for that to work. For example a "fully portable" instance of rustdesk is one way. If the user is admin then we need to cleanup any: `RustDesk.toml` and `RustDesk2.toml`


Cleanup is advanced and I believe not necessary. If config stuff is saved at the same place where standalone EXE is, and that is out of system folders then no special privileges are needed to create or delete them.

I am pretty sure, if resdeks.exe standalone does not save config into %appdata% directory but at the directory where exe is, problems would be resolved.

Change in code should be fairly simple as described in https://github.com/rustdesk/rustdesk/discussions/2127

**Comment by taveren** [comment]
Is there a way to to solve same problem in mac client? Solution with file naming with win. client looks strange but works fine. Maybe it's possible to do the same with mac version? The goal is to have one file for each platform, that can be downloaded with one click from own server and had integrated settings of this server.

**Comment by pedjas** [comment]
More issues with using file name for configuration parameters:

When link to download such file is sent via email or other service, link is not properlz recognized. Other side cannot just click it to download.

Antivirus programs scream on this. Some do scan and find it is not harmful, others simply block it recognizing spyware or other kind of malware.

**Reply by elico** [reply]
@pedjas I believe that there is a site which scans the files with a set of AV software.
As far as I can tell MS Defender and couple others are OK with it.
I would assume that some specific AV recognize rustdesk as spyware but it's another matter and should be opened in another issue.

**Comment by jmtatsch** [comment]
How can setting host and key be achieved parametrically for linux and Mac?
Can it be done in RustDesk.toml?

**Comment by coucoy** [comment]
Personally, for a network distributions, more in a corporate network or a true Self-hosted Network :smiley: 

i would suggest using DNS for automatic configuration:
using dnsmasq as configuration example:
srv-host=_hbbs._tcp,hbbs.example.com,21115
txt-record=hbbs_key,key=ABCDEFGED

**Reply by plittlefield** [reply]
That’s an interesting solution!

**Reply by labor4** [reply]
That's a PERFECT solution! (yes if DNSSEC)

**Reply by C-Duv** [reply]
So, natively/out-of-the-box, RustDesk will try to resolve `_hbbs._tcp.{local_domain_name}` and `hbbs_key.{local_domain_name}`?

**Reply by riccardospeggiorin-centropaghe** [reply]
@coucoy Is this working on a public domain too? That would be great!

**Comment by paulsommer** [comment]
I don't understand why I have to enter the server public-key at all.
Why don't the client and server exchange their public keys automatically on connection like other products do. Isn't that the most obvious way to go?

btw. How to find the key for the public servers?

**Reply by mpember** [reply]
Without the client storing the public key, there is no way of identitying if a MITM attack has caused a false key to be returned.  The other systems that offer public key exchange have a trusted method of exchanging keys via an intermediate server.

**Reply by paulsommer** [reply]
Of course you are right. Silly me :-) Thanks for the clarification.

**Comment by SlalomJohn** [comment]
/etc/systemd/system/rustdeskrelay.service
```
[Unit]
Description=Rustdesk Relay Server
[Service]
Environment="KEY_PRIV=private_key_text"
Environment="KEY_PUB=pub_key_text"
Type=simple
LimitNOFILE=1000000
ExecStart=/opt/rustdesk/hbbr -k _
WorkingDirectory=/opt/rustdesk/
User=root
Group=root
Restart=always
StandardOutput=append:/var/log/rustdesk/relayserver.log
StandardError=append:/var/log/rustdesk/relayserver.error
# Restart service after 10 seconds if node service crashes
RestartSec=10
[Install]
WantedBy=multi-user.target
```

/etc/systemd/system/rustdesksignal.service
```
[Unit]
Description=Rustdesk Signal Server
[Service]
Environment="ALWAYS_USE_RELAY=Y"
Environment="RELAY_SERVERS=your_server_ip_or_name_where_is_running_hhbr"
Environment="KEY_PRIV=private_key_text"
Environment="KEY_PUB=pub_key_text"

Type=simple
LimitNOFILE=1000000
ExecStart=/opt/rustdesk/hbbs -r your_server_ip_or_name_where_is_running_hhbr -k _
WorkingDirectory=/opt/rustdesk/
User=root
Group=root
Restart=always
StandardOutput=append:/var/log/rustdesk/signalserver.log
StandardError=append:/var/log/rustdesk/signalserver.error
# Restart service after 10 seconds if node service crashes
RestartSec=10
[Install]
WantedBy=multi-user.target
```

Client binary .EXE rename as
`rustdesk-host=YOUR_SERVER_IP_OR_NAME,key=PUB_KEY_TEXT.exe`

sample
`rustdesk-host=rd.microsoft.com,key=N445gfAAbBbbds6p0tIs=.exe`

Text keys public and private got from /opt/rustdesk
It's names as
private - id_xxxxxxxx
public - id_xxxxxxxx.pub

This is 100% working solution on current day
Server - Debian 11
Client's - Windows 10x64 / 11x64

**Reply by mpember** [reply]
> This is 100% working solution on current day

What problem are you offering this as a solution to?  It appears unrelated to the issue being discussed.

**Reply by SlalomJohn** [reply]
> > This is 100% working solution on current day
> What problem are you offering this as a solution to? It appears unrelated to the issue being discussed.

Situation:
self server on debian with real IP with key auth only
client in localnet to client on localnet - connect sucess
client in localnet to client on internet - connect is broken
client on internet to client in localnet - connect sucess

after set this settings all problem is away

P.S. no active firewall is present to outgoing traffic...

**Reply by mpember** [reply]
> Situation

That is not what this discussion is about.  As the OP states in their opening sentence, we are fully aware of the filename-based method of sharing the key with a client.  This discussion was started as a way of discussing possible alternatives and methods that are available for non-Windows clients.

**Comment by Poordeveloper** [comment, marked as answer]
https://rustdesk.com/docs/en/self-host/client-configuration/

**Reply by bigretromike** [reply]
looks like the generator need to be reverse engineering as you are unable to generate config script without paywall

**Reply by C-Duv** [reply]
I don't' think it's really a configuration script: I see it as an .app generator (for macOS at least)

Having compared public 1.3.2 release with what the client generator generates: it "simply" adds an `custom.txt` file in the resulting`my-application.app/Contents/Resources/custom.txt`.

This `custom.txt` file partially Base64 encoded ("partially" because I still gibberish characters) and holds the configuration as a JSON (+ other stuff I can't decode).

Either the config string pictured in the Wiki is not available anymore either it's not available for macOS clients.

**Reply by bigretromike** [reply]
generating that string config would be enough

**Reply by jameskimmel** [reply]
IMHO a very bad thing to paywall. 

I get paywalling certain stuff, and I get that you have to make a living. 
But naming Rustdesk as "OpenSource Remote Desktop with Self-Hosted Server Solutions" seems a little bit scammy, when you paywall the only reasonable way to self host. 

I don't have anything against paying or donating for OSS. 
But for me personally this isn't the way to do it and leaves a sour taste in my mouth. 
This makes me rather donate for the client generator here https://rdgen.crayoneater.org
or pay for a not that much more expensive Teamviewer license.  

I get that everyone is different in that regard. I just wanted to show you some insight on how you might sell more licenses with that strategy, but also miss out on some others. And maybe from a business perspective your way is even the more profitable way.

**Comment by AndiDad** [comment]
Passing the server ID and key via filename no longer works for me since version 1.3.9.

**Reply by AndiDad** [reply]
> Passing the server ID and key via filename no longer works for me since version 1.3.9.

The data will be transferred after installation. Simply starting the .exe file is not sufficient.

**Comment by afu8888** [comment]
您的来信已收到，我会尽快阅读！
如有急事，请致电本人！
